### PR TITLE
improve visualization for MoveRelative stage

### PR DESCRIPTION
**Pull request by j-kuehn**
Implementation for this [TODO](https://github.com/ros-planning/moveit_task_constructor/blob/master/core/src/stages/move_relative.cpp#L273)

The `MoveRelative` stage can now be visualized as a red-green arrow, where the red part shows the invalid part and the green part shows the valid part of the plan. The construction of the arrow is different for forward and backward propagators.

For forward propagators, it looks like this:

https://user-images.githubusercontent.com/57101356/114857654-cfd8d680-9de8-11eb-8a97-ac9b20e298d0.mp4

and for backward propagators like this:

https://user-images.githubusercontent.com/57101356/114857852-0d3d6400-9de9-11eb-9021-32ebb64ec019.mp4

Unfortunately, for forward propagators with only short invalid parts, the arrow tip looks weird: 

https://user-images.githubusercontent.com/57101356/114857959-2ba35f80-9de9-11eb-8e70-b97b590edb32.mp4

I couldn't think of a reasonably easy way around this (except for removing the tip entirely), so I left it like this.

I changed the `rviz_marker_tools::makeArrow` signature to explicitly include the diameter, so that the diameter is not a function of the length anymore. I also removed the default value, as none of the other `make*` functions have one and I found no reference of the default being used.

Also, I was curious why the plan should only be visualized when the planned trajectory is longer than `1e-3`. I left it in there for now, but think this restriction could also be removed.

### CODE REVIEW

File: rviz_marker_tools/include/rviz_marker_tools/marker_creation.h [hunk 1]
```diff
@@ -66,7 +66,8 @@ inline visualization_msgs::Marker& makeMesh(visualization_msgs::Marker& m, const
 }
 
 /// create an arrow along x-axis
-visualization_msgs::Marker& makeArrow(visualization_msgs::Marker& m, double scale = 1.0, bool tip_at_origin = false);
+visualization_msgs::Marker& makeArrow(visualization_msgs::Marker& m, double diameter, double length,
+                                      bool tip_at_origin = false);
```
Comment (rhaschke) [inline rviz_marker_tools/include/rviz_marker_tools/marker_creation.h:70]: Please keep the old API and forward to the new function as `makeArrow(m, 0.1*scale, scale, tip_at_origin)`

**Review by rhaschke — COMMENTED** [review]
The videos look good. Did you also consider the case of a pure rotation? In this case, the _linear_ arrow would be zero-length, although the `distance` would be non-zero. This was protected via the `1e-3` threshold before. Now, you seem to mix these two.
To improve on too short arrowheads: What about using an arrow tip without any shaft (of the same color). The shaft would be the cylinder of the other color.

**Comment by j-kuehn** [comment]
I had not considered the case of pure rotation, thanks for the hint. But since I left the protection in there, it still works as expected: no visualization for a pure rotation, the `distance` is only used if `linear_norm > 1e-3`. Wouldn't it make more sense to use the `use_rotation_distance` variable for the decision to visualize or not, though? It also decides whether the `distance` is calculated as linear or angular motion, so it seems more consistent to me than the somewhat arbitrary `1e-3`.

Only using an arrowhead without a shaft for short invalid parts would require defining the arrow with start/end points instead of position/orientation (see [here](http://wiki.ros.org/rviz/DisplayTypes/Marker#Arrow_.28ARROW.3D0.29)). With the available interface for creating markers, there would always be a 'cut' in the visualization at the point where the red invalid part switches from only being an arrowhead to having the normal arrow proportions.

TL;DR Imho the tradeoff between the additional code complexity a case distinction would bring and the improved visualization for short invalid parts is not worth it and I'd stick with the short arrowheads for the sake of simplicity. If you think the benefit justifies an increased code complexity, I'd be happy to try and implement it, though.

**Comment by rhaschke** [comment]
Yes, the distance definitely should consider the linear norm only. Did you try a real twist composed of rotation + translation along some common axis?
Regarding the arrowheads: I agree that extra complexity should be avoided. However, what about a slightly thinner arrow surrounded by cylinder occluded the longer arrow shaft?

File: core/src/stages/move_relative.cpp [hunk 2]
```diff
@@ -104,6 +104,59 @@ static bool getJointStateFromOffset(const boost::any& direction, const moveit::c
 	return false;
 }
 
+static void shapeMarker(visualization_msgs::Marker& m, uint32_t shape, rviz_marker_tools::Color color_id,
+                        const Eigen::Vector3d& pos, const Eigen::Quaterniond& quat, double width, double length) {
+	rviz_marker_tools::setColor(m.color, color_id);
```
Comment (v4hn) [inline core/src/stages/move_relative.cpp:109]: Why do you set color before "making" the marker? I had to check to see the color's not actually reset below.

**Review by v4hn — COMMENTED** [review]
> Wouldn't it make more sense to use linear_norm > 0 for the decision to visualize or not, though?

I agree. Especially because [the same check is already used in the class](https://github.com/ros-planning/moveit_task_constructor/blob/master/core/src/stages/move_relative.cpp#L177) to differentiate between pure rotational "mixed" modes.

> Imho the tradeoff between the additional code complexity a case distinction would bring and the improved visualization for short invalid parts is not worth it and I'd stick with the short arrowheads for the sake of simplicity

I'm unsure about this one. It would be nice to get a "normal" full arrow-head [with factor 0.23 of the whole arrow](https://github.com/ros-visualization/rviz/blob/noetic-devel/src/rviz/default_plugin/markers/arrow_marker.cpp#L105) of one color, even if the actual unreachable part of the trajectory is very short.
But then again, it's more accurate to show the translation that really failed and not exaggerate.

**Comment by j-kuehn** [comment]
The arrow creation is now implemented with passing start and end points. I first reverted to the old API, then overloaded `makeArrow` to allow this alternative of defining the arrow.

The benefit of this is that the `scale.z` now allows to set the length of the arrowhead. Accordingly, I set this to `0.23 * linear_norm` to keep the default proportions of arrows:

backward-invalid
forward-invalid

If the invalid part (or the valid part for backward direction) is too short to have the default proportions, [this](https://github.com/ros-visualization/rviz/blob/noetic-devel/src/rviz/default_plugin/markers/arrow_marker.cpp#L110) ensures that the arrow only consists of an arrowhead without a shaft:
forward-invalid-short

For long valid parts and really short invalid parts it looks like this now:

https://user-images.githubusercontent.com/57101356/116075855-c3c80100-a693-11eb-988c-8101a5e6504a.mp4

Backward direction accordingly:
backward-invalid-short

Imho this visualization is good enough.

> Did you try a real twist composed of rotation + translation along some common axis?

I did test multiple twist messages and this works as expected. If the target direction only consists of a rotation, no visualization is computed. If the target direction consists of both a rotation and translation, the visualization is computed along the translational part.

File: core/src/stages/move_relative.cpp [hunk 3]
```diff
@@ -104,6 +104,59 @@ static bool getJointStateFromOffset(const boost::any& direction, const moveit::c
...
+	} else if (shape == visualization_msgs::Marker::CYLINDER) {
+		rviz_marker_tools::makeCylinder(m, width, length);
+	}
+	tf::pointEigenToMsg(pos, m.pose.position);
+	tf::quaternionEigenToMsg(quat, m.pose.orientation);
+}
+
+static void visualizePlan(Interface::Direction dir, bool success, SubTrajectory& solution, const std::string& ns,
+                          const std::string& frame_id, Eigen::Quaterniond& quat, Eigen::Vector3d& pos,
+                          double linear_norm, double distance) {
```
[the remark this answers is not in the document]
Comment (j-kuehn) [reply core/src/stages/move_relative.cpp]: I am now passing the `reached_pose` and calculate the required rotation and translations inside the method. I am also passing `link_pose` to get the translation before planning. I could also compute this from the `reached_pose` (or go from `link_pose` to `reached_pose`) inside the method and have a parameter less, but it seemed more robust and concise to me this way.

File: core/src/stages/move_relative.cpp [hunk 4]
```diff
@@ -104,6 +104,63 @@ static bool getJointStateFromOffset(const boost::any& direction, const moveit::c
...
+	auto linear_norm = linear.norm();
+
+	// rotation of the target direction and for the cylidner marker
+	auto quat_target = Eigen::Quaterniond::FromTwoVectors(Eigen::Vector3d::UnitX(), linear);
+	auto quat_cylinder = quat_target * Eigen::AngleAxisd(0.5 * M_PI, Eigen::Vector3d::UnitY());
+
+	// link position before planning; reached link position after planning; target link position
+	Eigen::Vector3d pos_link = link_pose.translation();
+	Eigen::Vector3d pos_reached = reached_pose.translation();
+	Eigen::Vector3d pos_target = pos_reached + quat_target * Eigen::Vector3d(linear_norm - distance, 0, 0);
```
Comment (v4hn) [inline core/src/stages/move_relative.cpp:119]:
```suggestion
	auto& pos_link = link_pose.translation();
	auto& pos_reached = reached_pose.translation();
	auto pos_target = pos_reached + quat_target * Eigen::Vector3d(linear_norm - distance, 0, 0);
```
... is what I would write, because the explicit types will actually evaluate Eigen's templated lazy evaluation before it's actually needed (the auto-deduced types are not Vector3d, but some `EigenBase<...>` datatypes). It won't make a difference here though, so feel free to leave it as is. See for example [here](https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html) for something to start reading about the subject.

File: rviz_marker_tools/src/marker_creation.cpp [hunk 5]
```diff
@@ -276,6 +276,22 @@ vm::Marker& makeMesh(vm::Marker& m, const std::string& filename, double sx, doub
 	return m;
 }
 
+vm::Marker& makeArrow(vm::Marker& m, const Eigen::Vector3d& start_point, const Eigen::Vector3d& end_point, double diameter, double head_length) {
+	// scale.y is set according to default proportions in rviz/default_plugin/markers/arrow_marker.cpp#L61
+	// for the default head_length=0, the head length will keep the default proportion defined in arrow_marker.cpp#L106
+	m.scale.x = diameter;
+	m.scale.y = 2 * diameter;
+	m.scale.z = head_length;
+	prepareMarker(m, vm::Marker::ARROW);
```
Comment (v4hn) [inline rviz_marker_tools/src/marker_creation.cpp:285]:
```suggestion
	prepareMarker(m, vm::Marker::ARROW);
	// scale.y is set according to default proportions in rviz/default_plugin/markers/arrow_marker.cpp#L61
	// for the default head_length=0, the head length will keep the default proportion defined in arrow_marker.cpp#L106
	m.scale.x = diameter;
	m.scale.y = 2 * diameter;
	m.scale.z = head_length;

Comment (j-kuehn) [reply rviz_marker_tools/src/marker_creation.cpp:285]: I put `prepareMarker` after the `scale` assignments, because it ensures a valid scale. Currently, this is not necessary, as `diameter` and `head_length` are derived from `linear_norm`, which is verified to be non-zero for the visualization.

However, if this should ever change in the future, I thought it's better to have that scale check, especially since all other methods call `prepareMarker` after the `scale` assignments to ensure a valid scale.

File: core/src/stages/move_relative.cpp (hunk 4 above)
Comment (j-kuehn) [reply core/src/stages/move_relative.cpp:119]: I am unsure why this doesn't work, but I tried it with `auto` and `auto&` before and just did again and it breaks the visualization. Since I simply pass the variables to `tf::pointEigenToMsg`, it has to be something in there, but I didn't do any further troubleshooting.

**Review by v4hn — APPROVED** [review]

File: rviz_marker_tools/src/marker_creation.cpp [hunk 6]
```diff
@@ -276,6 +276,20 @@ vm::Marker& makeMesh(vm::Marker& m, const std::string& filename, double sx, doub
...
+                      double diameter, double head_length) {
+	// scale.y is set according to default proportions in rviz/default_plugin/markers/arrow_marker.cpp#L61
+	// for the default head_length=0, the head length will keep the default proportion defined in arrow_marker.cpp#L106
+	m.scale.x = diameter;
+	m.scale.y = 2 * diameter;
+	m.scale.z = head_length;
+	prepareMarker(m, vm::Marker::ARROW);
+	m.points.resize(2);
+	tf::pointEigenToMsg(start_point, m.points[0]);
+	tf::pointEigenToMsg(end_point, m.points[1]);
```
Comment (rhaschke) [inline rviz_marker_tools/src/marker_creation.cpp:289]: It would have been better to directly use tf2 functions. See https://github.com/ros-planning/moveit_task_constructor/issues/78#issuecomment-826788870. But I guess, Michael will give you this as a next  task anyway :wink:

**Review by rhaschke — APPROVED** [review]

**Comment by rhaschke** [comment]
Very nice solution and visualization!
